crates/valence_command/src/arg_parser.rs

```diff
@@ -0,0 +1,970 @@
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::ResourceLocation
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct Function;
+
+impl CommandArg for Function {
+    type Result = String;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(string)
+    }
+
+    fn display() -> Parser {
+        Parser::Function
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum EntityAnchor {
+    #[default]
+    Eyes,
+    Feet,
+}
+
+impl CommandArg for EntityAnchor {
+    type Result = EntityAnchor;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        Ok(match string.as_str() {
+            "eyes" => EntityAnchor::Eyes,
+            "feet" => EntityAnchor::Feet,
+            _ => {
+                return Err(CommandArgParseError::InvalidArgument(
+                    "entity_anchor".to_string(),
+                    string,
+                ))
+            }
+        })
+    }
+
+    fn display() -> Parser {
+        Parser::EntityAnchor
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub struct IntRange {
+    min: i32,
+    max: i32,
+}
+
+impl CommandArg for IntRange {
+    type Result = IntRange;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        let mut split = string.split("..");
+        let min = split.next().unwrap().parse::<i32>().map_err(|_| {
+            CommandArgParseError::InvalidArgument("int_range max".to_string(), string.clone())
+        })?;
+        let max = split.next().unwrap().parse::<i32>().map_err(|_| {
+            CommandArgParseError::InvalidArgument("int_range min".to_string(), string.clone())
+        })?;
+
+        Ok(IntRange { min, max })
+    }
+
+    fn len() -> ArgLen {
+        ArgLen::Exact(2)
+    }
+
+    fn display() -> Parser {
+        Parser::IntRange
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Default)]
+pub struct FloatRange {
+    min: f32,
+    max: f32,
+}
+
+impl CommandArg for FloatRange {
+    type Result = FloatRange;
+
+    fn arg_from_string(string: String) -> Result<Self::Result, CommandArgParseError> {
+        let mut split = string.split("..");
+        let min = split.next().unwrap().parse::<f32>().map_err(|_| {
+            CommandArgParseError::InvalidArgument("float_range max".to_string(), string.clone())
+        })?;
+        let max = split.next().unwrap().parse::<f32>().map_err(|_| {
+            CommandArgParseError::InvalidArgument("float_range min".to_string(), string.clone())
+        })?;
+
+        Ok(FloatRange { min, max })
+    }
+
+    fn len() -> ArgLen {
+        ArgLen::Exact(2)
+    }
+
+    fn display() -> Parser {
+        Parser::FloatRange
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum Dimension {
+    #[default]
+    Overworld,
+    Nether,
+    End,
+}
```

Comment:
Are these three dimensions hard coded, or does the client read from the dimension registry? If it's the ladder then it should probably be an arbitrary string.